File: app/src/main/java/com/hjc/reader/viewmodel/menu/CollectLinkViewModel.java
package com.hjc.reader.viewmodel.menu;


import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.blankj.utilcode.util.ToastUtils;
import com.hjc.baselib.http.RxSchedulers;
import com.hjc.baselib.http.observer.BaseProgressObserver;
import com.hjc.baselib.viewmodel.CommonViewModel;
import com.hjc.reader.bean.response.CollectArticleBean;
import com.hjc.reader.bean.response.CollectLinkBean;
import com.hjc.reader.http.RetrofitHelper;
import com.hjc.reader.http.observer.BasePageObserver;

import java.util.List;

public class CollectLinkViewModel extends CommonViewModel {

    // 收藏地址列表数据
    private MutableLiveData<List<CollectLinkBean.DataBean>> collectLinkLiveData = new MutableLiveData<>();

    private MutableLiveData<Integer> positionLiveData = new MutableLiveData<>();

    public CollectLinkViewModel(@NonNull Application application) {
        super(application);
    }

    /**
     * 获取收藏地址列表数据
     *
     * @param isFirst 是否第一次加载
     */
    public void loadCollectLinkList(boolean isFirst) {
        RetrofitHelper.getInstance().getWanAndroidService()
                .getLinkList()
                .compose(RxSchedulers.ioToMain())
                .subscribe(new BasePageObserver<CollectLinkBean>(this, isFirst) {
                    @Override
                    public void onSuccess(CollectLinkBean result) {
                        List<CollectLinkBean.DataBean> dataList = result.getData();
                        if (dataList != null && dataList.size() > 0) {
                            collectLinkLiveData.setValue(dataList);
                        } else {
                            collectLinkLiveData.setValue(null);
                            showEmpty();
                            ToastUtils.showShort("暂无收藏,快去收藏吧");
                        }
                    }
                });
    }

    /**
     * 删除收藏的网址
     *
     * @param id 网址id
     */
    public void deleteLink(int id, int position) {
        RetrofitHelper.getInstance().getWanAndroidService()
                .deleteLink(id)
                .compose(RxSchedulers.ioToMain())
                .subscribe(new BaseProgressObserver<CollectArticleBean>(this) {
                    @Override
                    public void onSuccess(CollectArticleBean result) {
                        if (result.getErrorCode() == 0) {
                            positionLiveData.setValue(position);
                            ToastUtils.showShort("删除网址成功");
                        } else {
                            ToastUtils.showShort(result.getErrorMsg());
                        }
                    }
                });
    }


    public MutableLiveData<List<CollectLinkBean.DataBean>> getCollectLinkLiveData() {
        return collectLinkLiveData;
    }

    public MutableLiveData<Integer> getPositionLiveData() {
        return positionLiveData;
    }

}

File: module-main/src/main/java/com/hjc/learn/main/viewmodel/collect/CollectLinkViewModel.kt
package com.hjc.learn.main.viewmodel.collect

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.blankj.utilcode.util.ToastUtils
import com.hjc.library_common.viewmodel.KotlinViewModel
import com.hjc.library_net.RetrofitClient
import com.hjc.library_net.model.WanCollectLinkBean
import java.net.SocketTimeoutException
import java.net.UnknownHostException

/**
 * @Author: HJC
 * @Date: 2021/3/3 11:00
 * @Description: 收藏网址ViewModel
 */
class CollectLinkViewModel(application: Application) : KotlinViewModel(application) {

    // 收藏地址列表数据
    val collectLinkLiveData = MutableLiveData<MutableList<WanCollectLinkBean>>()

    // 刷新数据
    val refreshData = MutableLiveData<Boolean>()

    // 取消收藏位置数据
    val positionLiveData = MutableLiveData<Int>()


    /**
     * 获取收藏地址列表数据
     *
     * @param isFirst 是否第一次加载
     */
    fun loadCollectLinkList(isFirst: Boolean) {
        launchWan({ RetrofitClient.getApiService1().getLinkList() }, { result ->
            refreshData.value = true

            result?.let {
                if (it.size > 0) {
                    collectLinkLiveData.value = it
                    showContent()
                } else {
                    showEmpty()
                }
            }
        }, { e ->
            refreshData.value = true

            if (e is UnknownHostException || e is SocketTimeoutException) {
                showError()
            } else {
                showTimeout()
            }
        }, isShowProgress = isFirst)
    }

    /**
     * 删除收藏的网址
     *
     * @param id 网址id
     */
    fun deleteLink(id: Int, position: Int) {
        launchWan({ RetrofitClient.getApiService1().deleteLink(id) }, {
            positionLiveData.value = position
            ToastUtils.showShort("删除网址成功")
        }, isShowLoading = true)
    }

}